### DISCUSSION TITLE
Recommandation to augment `$fetch` in a module

### DISCUSSION THREAD

**Discussion by Barbapapazes** [nuxt/nuxt#29067] [Questions]
### Describe the feature

Hello 👋,

_This is both a question and a potential feature request._

I'm currently developing a Nuxt module and was wondering if there's a way to extend `$fetch`. Specifically, my module needs to add a header to each request.

```vue
<script lang="ts" setup>
async function retrieveData() {
  await $fetch('/api/data', {
    method: 'POST',
  })
}
</script>
```

This request requires a custom token to be authorized by the server. While I could manually set the `x-token` header, this solution works well within the application but may cause conflicts with other modules.

For example, the [nuxt-auth-utils](https://github.com/atinux/nuxt-auth-utils) module by @atinux clears the session by making a `DELETE` request. The end-user simply calls [the `clear` function](https://github.com/atinux/nuxt-auth-utils/blob/main/src/runtime/app/composables/session.ts#L37). Without a way to extend `$fetch`, I would need to rewrite this function in my codebase and modify it to include my custom header.

I found a workaround by creating a Nuxt plugin that wraps the fetch function, allowing me to tweak it as needed.

```ts
export default defineNuxtPlugin({
  name: 'x-token',
  setup() {
    if (import.meta.env.SSR) {
      return
    }

    const originalFetch = $fetch
    // eslint-disable-next-line no-global-assign
    $fetch = async (url, options) => {
      console.log('fetch', url, options)
      const token = document.cookie.match(/token=([^;]+)/)?.[1]

      const isSameOrigin = !url.startsWith('http') || url.startsWith(location.origin)

      if (token && isSameOrigin) {
        options = options ?? {}
        options.headers ??= {}
        options.headers['x-token'] = token
      }

      return originalFetch(url, options)
    }
  },
})
```

### Additional information

- [ ] Would you be willing to help implement this feature?
- [ ] Could this feature be implemented as a module?

### Final checks

- [X] Read the [contribution guide](https://nuxt.com/docs/community/contribution).
- [X] Check existing [discussions](https://github.com/nuxt/nuxt/discussions) and [issues](https://github.com/nuxt/nuxt/issues).

**Comment by pi0** [comment]
Some notes [here](https://github.com/unjs/ofetch/issues/431#issuecomment-2306946446) about ofetch and design.

I think you can make a feature request directly to `nuxt-auth-utils` to allow application wide headers for `useUserSession().fetch` composable via config or plugin. (however a word from someone maintained popular axios + auth modules for nuxt 2: application wide mutations are not safe. it will be much much safer now since only such change would affect `useUserSession()` but i won't be surprised if any edge cases happen with that too)

**Comment by Barbapapazes** [comment]
Thanks for this advice.

I have created a utils but it's very hard to infer the proper type for the options.

```ts
import type { NitroFetchOptions, NitroFetchRequest } from 'nitropack'

export function withToken<R extends NitroFetchRequest = NitroFetchRequest, O extends NitroFetchOptions<R> = NitroFetchOptions<R>>(options: O): O {
	return options // This is where I will inject the token
}
```

Usage:

```ts
  await $fetch('/api/data', withToken({
    method: 'post',

    headers: {
      'Content-Type': 'application/json',
    },
  }))
```

But because the options needs the `url`, the dev loose some types within my util. 

Is there a better way to do it?

**Comment by danielroe** [comment]
Instead of wrapping just the options, it's better for inference to wrap the entire `$fetch` function, as long as you are simply adding a header.
